### PR TITLE
Used a conditional type for Sinon's stub class typings 

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
Continues https://github.com/DefinitelyTyped/DefinitelyTyped/pull/24399.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/9f4c75126167d0d8af759f58405d53d983e94ad0/types/sinon/index.d.ts#L570
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Review by pimterry — APPROVED** [review]

**Comment by pavelmickevic** [comment]
This is a breaking change. Started seeing errors in output: 
> node_modules\@types\sinon\index.d.ts
> TS1005: ';' expected.
> ...
> Property 'spy' does not exist on type 'typeof Sinon'.
> ...
> TS2339: Property 'sandbox' does not exist on type 'typeof Sinon'.

**Comment by 43081j** [comment]
@pavelmickevic this is because he has updated the code to only be compatible with typescript 2.8.

@andy-ms didn't we say in #25491 that its best not to merge the 2.8-only changes until it is more widely supported/used?

**Comment by 43081j** [comment]
@JoshuaKGoldberg the 5.x PR had these conditional types in it originally (taken from your PR in fact) and we had decided against merging them until 2.8 is wider spread. The problem is, it is not as simple as updating your project to use TS 2.8, because you may depend on other things (including your own code) which _only_ work on <2.8.

The PR is great and the change is much needed (original types were loose and incorrect), though.

So basically:

**If you do not yet use 2.8, do not update your sinon types (ever)**.

**Comment by JoshuaKGoldberg** [comment]
Heh, it's been reverted. I'll file a separate issue for now allowing new-TS-version specific features to go into popular packages quite so quickly.

**Comment by jonnocraig** [comment]
@JoshuaKGoldberg thanks for the quick update!

**Comment by JoshuaKGoldberg** [comment]
I didn't do anything, but my pleasure ;)
